airflow/models/deadline.py

```diff
@@ -0,0 +1,90 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+from datetime import datetime
+from typing import TYPE_CHECKING
+
+import sqlalchemy_jsonfield
+from sqlalchemy import Column, DateTime, Index, Integer, String
+
+from airflow.models.base import Base, StringID
+from airflow.settings import json
+from airflow.utils.log.logging_mixin import LoggingMixin
+from airflow.utils.session import NEW_SESSION, provide_session
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+
+class Deadline(Base, LoggingMixin):
+    """A Deadline is a 'need-by' date which triggers a callback if the provided time has passed."""
+
+    __tablename__ = "deadline"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+
+    # If the Deadline Alert is for a DAG, store the DAG ID and Run ID from the dag_run.
+    dag_id = Column(StringID())
+    run_id = Column(StringID())
```

Comment:
Later in the project I guess you'll want to join this table with other tables such as DAG/DagRun. Without foreign keys you might have issues doing it. I'd recommend adding foreign keys on these columns. You can have foreign keys without having the `onDelete=CASCADE` if this is something you do not want.

Plus, `run_id` is not a primary key of `DagRun`, so referencing a dag run with a column that is not a primary is off. I'd strongly recommend using the column `id` from `DagRun`

Reply:
Let's see if I end up needing it first?

Reply:
looks like i had the same thought as vincent

Reply:
Resolving this conversation to consolidate the topic in [this](https://github.com/apache/airflow/pull/44712#discussion_r1873960150) comment thread.